packages/cli/commands/import/workflow.ts

```diff
@@ -93,41 +117,94 @@ export class ImportWorkflowsCommand extends Command {
 						workflow.nodes.forEach((node: INode) => {
-							this.transformCredentials(node, credentialsEntities);
+							this.transformCredentials(node, credentials);
 						});
 					}
-					// eslint-disable-next-line no-await-in-loop, @typescript-eslint/no-non-null-assertion
-					await Db.collections.Workflow!.save(workflow);
+					await this.storeWorkflow(workflow, user);
 				}
 			} else {
-				const fileContents = JSON.parse(fs.readFileSync(flags.input, { encoding: 'utf8' }));
+				const workflows = JSON.parse(fs.readFileSync(flags.input, { encoding: 'utf8' }));
 
-				if (!Array.isArray(fileContents)) {
-					throw new Error(`File does not seem to contain workflows.`);
+				if (!Array.isArray(workflows)) {
+					throw new Error(
+						'File does not seem to contain workflows. Make sure the workflows are contained in an array.',
+					);
 				}
 
-				for (i = 0; i < fileContents.length; i++) {
-					if (credentialsEntities.length > 0) {
-						// eslint-disable-next-line
-						fileContents[i].nodes.forEach((node: INode) => {
-							this.transformCredentials(node, credentialsEntities);
+				for (i = 0; i < workflows.length; i++) {
+					if (credentials.length > 0) {
+						workflows[i].nodes.forEach((node: INode) => {
+							this.transformCredentials(node, credentials);
 						});
 					}
-					// eslint-disable-next-line no-await-in-loop, @typescript-eslint/no-non-null-assertion
-					await Db.collections.Workflow!.save(fileContents[i]);
+					await this.storeWorkflow(workflows[i], user);
 				}
 			}
 
 			console.info(`Successfully imported ${i} ${i === 1 ? 'workflow.' : 'workflows.'}`);
-			process.exit(0);
+			process.exit();
 		} catch (error) {
-			console.error('An error occurred while exporting workflows. See log messages for details.');
-			// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-			logger.error(error.message);
+			console.error('An error occurred while importing workflows. See log messages for details.');
+			if (error instanceof Error) logger.error(error.message);
 			this.exit(1);
 		}
 	}
+
+	private async initOwnerWorkflowRole() {
+		const ownerWorkflowRole = await Db.collections.Role!.findOne({
+			where: { name: 'owner', scope: 'workflow' },
+		});
+
+		if (!ownerWorkflowRole) {
+			throw new Error(`Owner workflow role not found. ${FIX_INSTRUCTION}`);
+		}
+
+		this.ownerWorkflowRole = ownerWorkflowRole;
+	}
+
+	private async storeWorkflow(workflow: object, user: User) {
+		await getConnection().transaction(async (transactionManager) => {
+			const newWorkflow = new WorkflowEntity();
+
+			Object.assign(newWorkflow, workflow);
+
+			const savedWorkflow = await transactionManager.save<WorkflowEntity>(newWorkflow);
+
+			const newSharedWorkflow = new SharedWorkflow();
+
+			Object.assign(newSharedWorkflow, {
+				workflow: savedWorkflow,
+				user,
+				role: this.ownerWorkflowRole,
+			});
+
+			await transactionManager.save<SharedWorkflow>(newSharedWorkflow);
+		});
```

Comment:
This syntax with initialization is required because `EntityManager` requires a full instance of the entity, whereas `Repository` does not. The only way to use your preferred syntax here is adjusting the model fields, which does not seem worth it.

```ts
// EntityManager
save<Entity>(entity: Entity, options?: SaveOptions): Promise<Entity>;

// Repository
save<T extends DeepPartial<Entity>>(entity: T, options?: SaveOptions): Promise<T & Entity>;
```

Reply:
One transaction per workflow or per import? Let me know what behavior we prefer.

Reply:
that's the question in optimal case it would be per import, because then we can ensure that you simply run the command again so we use the all or nothing principle 👍🏼